Update layout.md (#6868)
taichi-dev/taichi: docs/lang/articles/basic/layout.md

After familiarizing yourself with the basics of Taichi [Fields](./field.md), this article helps you one step further by explaining the underlying memory layout that is essential to write high-performance Taichi programs. In particular, we present how to organize an efficient data layout and how to manage memory occupancy.

> ## Organizing an Efficient Data Layout

Reviewer: ```suggestion
## Organizing an efficient data layout
```

The above 2D array built with nested `dense` statements is *not* equivalent to the 2D array built with `ti.field`.
Although both statements result in a 2D array of the same shape, they have
different layers of `SNodeTree`, as shown below:

> The following snippet has two `SNodeTree` layers below the root;

Reviewer: ```suggestion
The following snippet has two `SNodeTree` layers below the root:
```

```python
x = ti.field(ti.f32)
ti.root.dense(ti.i, 3).dense(ti.j, 4).place(x)
```

> The following snippet only has one `SNodeTree` below the root

Reviewer: ```suggestion
The following snippet only has one `SNodeTree` below the root:
```

```python
x = ti.field(ti.f32)
ti.root.dense(ti.ij, (3, 4)).place(x)

# or equivalently
x = ti.field(ti.f32, shape=(3,4))
```

> See the sketch below to see a visualization that illustrates the difference between these layers:

Reviewer: ```suggestion
See the sketch below for a visualization that illustrates the difference between these layers:
```

By nesting multiple statements, we can construct a field with higher dimensions.
<!-- haidong: how far can we go? how many default axis exist? -->

In order to traverse the nested statements, you can use a `for` loop:

```python {1}
for i, j in A:
    A[i, j] += 1
```

The Taichi compiler is capable of automatically deducing the underlying data layout and applying a proper data access order. This is an advantage over most general-purpose programming languages where the access order has to be optimized manually.

> ### Row-Major versus Column-Major

Reviewer: ```suggestion
### Row-major versus column-major
```

> One important thing to note about memory addresses is that its space is linear. Without loss of generality, we omit the differences in data types and assume each data element has size 1. Moreover, we denote the starting memory address of a field as `base`, and the indexing formula for 1D Taichi fields is `base + i` for the `i-th` element.

Reviewer: ```suggestion
One important thing to note about memory addresses is that their space is linear. Without loss of generality, we omit the differences in data types and assume each data element has size 1. Moreover, we denote the starting memory address of a field as `base`, and the indexing formula for 1D Taichi fields is `base + i` for the `i-th` element.
```

### AoS versus SoA

> AoS means _array of structures_ and SoA means _structure of arrays_. Consider an RGB image with 4 pixels and 3 color channels, an AoS layout stores `RGBRGBRGBRGB` while an SoA layout stores `RRRRGGGGBBBB`. The selection of AoS or SoA layout largely depends on the access pattern to the field. Let's discuss a scenario to process large RGB images. The two layouts have the following arrangements in memory:

Reviewer: ```suggestion
AoS means _array of structures_ and SoA means _structure of arrays_. Consider an RGB image with four pixels and three color channels: an AoS layout stores `RGBRGBRGBRGB` while an SoA layout stores `RRRRGGGGBBBB`. The selection of an AoS or SoA layout largely depends on the access pattern to the field. Let's discuss a scenario to process large RGB images. The two layouts have the following arrangements in memory:
```
